### PR TITLE
Saas 14342 add indirect field to yarn

### PR DESCRIPTION
**Pull request by MorAlon1**

### CODE REVIEW

File: pkg/nodejs/yarn/parse.go [hunk 1]
```diff
@@ -122,7 +122,9 @@ func ignoreProtocol(protocol string) bool {
 	return false
 }
 
-func parseResults(patternIDs map[string]string, dependsOn map[string][]string) (deps []types.Dependency) {
+func parseResults(patternIDs map[string]string, dependsOn map[string][]string) (deps []types.Dependency, indirectMap map[string]bool) {
+	indirectMap = make(map[string]bool)
```
Comment (VaismanLior) [inline pkg/nodejs/yarn/parse.go:126]:
```suggestion
	uniqIndirectDeps = make(map[string]bool)
```

File: pkg/nodejs/yarn/testcase_deps_generator/libs_generator.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      result[libId].version = value.version;
+      result[libId].suffix = result[libId].suffix
+        ? `${result[libId].suffix}, ${key}`
+        : key;
+
+      result[libId].dependencies = value.dependencies ? value.dependencies : {};
+    }
+  }
+
+  //console.log(Object.entries(result).filter(([a,d]) => a.includes("parse5") || d.suffix.includes("parse5") ));
```
Comment (VaismanLior) [inline pkg/nodejs/yarn/testcase_deps_generator/libs_generator.js:95]: remove

File: pkg/nodejs/yarn/parse.go [hunk 3]
```diff
@@ -132,8 +134,15 @@ func parseResults(patternIDs map[string]string, dependsOn map[string][]string) (
 			ID:        libID,
 			DependsOn: depIDs,
 		})
+
+		// check if libId in indirectMap
```
Comment (VaismanLior) [inline pkg/nodejs/yarn/parse.go:138]: This comment is unrelated, fix it or remove

**Review by knqyf263 — COMMENTED** [review]
If I correctly understand what you're doing, it doesn't work.

For example, this project has a direct dependency `negotiator`.

```
$ cat package.json
{
  "name": "lock-parser",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "@yarnpkg/lockfile": "^1.1.0",
    "yaml": "^2.1.3",
    "express": "^4.18.2",
    "negotiator": "^0.6.0"
  }
}
```

But it is also a dependency of `accepts`.

```
accepts@~1.3.8:
  version "1.3.8"
  resolved "https://registry.yarnpkg.com/accepts/-/accepts-1.3.8.tgz#0bf0be125b67014adcb0b0921e62db7bffe16b2e"
  integrity sha512-PYAthTa2m2VKxuvSD3DPC/Gy+U+sOA1LAuT8mkmRuvw+NACSaeXEQ+NHcVF7rONl6qcaxV3Uuemwawk+7+SJLw==
  dependencies:
    mime-types "~2.1.34"
    negotiator "0.6.3"
```

Even if you find `negotiator` under `dependencies` it doesn't mean transitive dependencies.

**Comment by VaismanLior** [comment]
> If I correctly understand what you're doing, it doesn't work.
> 
> For example, this project has a direct dependency `negotiator`.
> 
> ```
> $ cat package.json
> {
>   "name": "lock-parser",
>   "version": "1.0.0",
>   "main": "index.js",
>   "license": "MIT",
>   "dependencies": {
>     "@yarnpkg/lockfile": "^1.1.0",
>     "yaml": "^2.1.3",
>     "express": "^4.18.2",
>     "negotiator": "^0.6.0"
>   }
> }
> ```
> 
> But it is also a dependency of `accepts`.
> 
> ```
> accepts@~1.3.8:
>   version "1.3.8"
>   resolved "https://registry.yarnpkg.com/accepts/-/accepts-1.3.8.tgz#0bf0be125b67014adcb0b0921e62db7bffe16b2e"
>   integrity sha512-PYAthTa2m2VKxuvSD3DPC/Gy+U+sOA1LAuT8mkmRuvw+NACSaeXEQ+NHcVF7rONl6qcaxV3Uuemwawk+7+SJLw==
>   dependencies:
>     mime-types "~2.1.34"
>     negotiator "0.6.3"
> ```
> 
> Even if you find `negotiator` under `dependencies` it doesn't mean transitive dependencies.

Isn't it the meaning of the `Indirect` field? I mean, a dependency can be direct and indirect at the same time.
Why don't we have a `Direct` field?
Using the lockfile only we could never know if a dependency is direct, but we can know if it's indirect.
